### PR TITLE
Feature/#66 pan while zoom

### LINKED ISSUE #66: Allow pan while pinching
# Current Behaviour
The focus point when pinching is fixed at the center of both touch points when the pinching gesture is started. This means the content can not be panned during the pinch gesture.

# Expected Behaviour
The focus point moves with the center of both touch points. 
When starting the pinch gesture the center is remembered and when moving both touch points in the same direction and speed (without changing their distance) the content should be panned without any zoom changes.

This allows to zoom more conveniently to a specific position instead of having to zoom to the desired level and adjusting view position afterwards using pan gesture.

### PR DESCRIPTION
**Pull request by markusressel**
fixes #66 

Wow so this took quite a while for me to get it to work but I'm quite happy with the results.
The hardest part was animating back to a non-overscrolled and non-overpinched state at the same time.
I use a "simulated" zoom state to calculate the pan fixes needed to be applied to get back into a valid state when the zoom is also changing. I will add a review comment to the mentioned line, please let me know if you know a better way to do this.

### CODE REVIEW

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 1]
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
                 val factor = detector.scaleFactor
                 val newZoom = zoom * factor
-                applyPinch(newZoom, mAbsTargetX, mAbsTargetY, true)
+
+                // apply the new zoom and pan values accordingly
+                applyZoomAndAbsolutePan(newZoom,
+                        mCurrentAbsTargetX, mCurrentAbsTargetY,
+                        allowOverScroll = true,
+                        allowOverPinch = true)
                 return true
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:851]: This is the stuff (above code) that allows pan and zoom to be applied simultaniously when using pinch gesture.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 2]
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+                    // select zoom pivot point based on what edge of the screen is currently overscrolled
+                    var zoomTargetX = calculateZoomPivotPoint(true, fixPanX)
+                    var zoomTargetY = calculateZoomPivotPoint(false, fixPanY)
+
+                    // calculate the new pan position
+                    var newPanX = panX + fixPanX
+                    var newPanY = panY + fixPanY
+
+                    if (newZoom.compareTo(zoom) != 0) {
+                        // we have overpinched. to calculate how much pan needs to be applied
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:894]: Here is the zoom "simulation" part. The problem I faced was that I couldn't just calculate pan fix with the current zoom and apply it at the same time as the new zoom level as the pan fix calculated by the `checkPanBounds` method is an absolute pan value that is always located at the top left of the view (I think?). Therefore this absolute position depends on the current zoom level - so the calculated fix is not applicable to the state after the overpinched zoom has been changed.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 3]
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+                        // recalculate pivot point based on the recalculated pan fix
+                        zoomTargetX = calculateZoomPivotPoint(true, fixPanX)
+                        zoomTargetY = calculateZoomPivotPoint(false, fixPanY)
+
+                        // recalculate new pan location using the simulated target zoom level
+                        newPanX = panX + fixPanX
+                        newPanY = panY + fixPanY
+
+                        // revert simulation
+                        applyZoom(oldZoom, true, true, zoomTargetX, zoomTargetY)
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Those `applyZoom` calls currently trigger listeners. I don't know if that's a problem but I think it should be avoided as it's just a simulation to calculate pan fix numbers?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 4]
```diff
@@ -1240,29 +1364,30 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+     * @param zoom        the new zoom
      * @param targetX        the target X in abs value
      * @param targetY        the target Y in abs value
      * @param allowOverPinch whether to overPinch
      */
-    private fun applyPinch(@Zoom newZoom: Float, @AbsolutePan targetX: Float, @AbsolutePan targetY: Float,
+    private fun applyPinch(@Zoom zoom: Float, @AbsolutePan targetX: Float, @AbsolutePan targetY: Float,
                            allowOverPinch: Boolean) {
-        var newZoom = newZoom
+        // TODO: this method is currently unused, should it be removed? The doc mentions buggy translations :S
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: You used this method to apply zoom **only** in the `PinchListener`.
I don't know if it is of any other value and should be kept as some kind of helper?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 5]
```diff
@@ -1044,25 +1155,23 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
-    private fun animateZoom(@Zoom newZoom: Float, allowOverPinch: Boolean) {
-        var newZoom = newZoom
-        newZoom = checkZoomBounds(newZoom, allowOverPinch)
+    private fun animateZoom(@Zoom zoom: Float, allowOverPinch: Boolean) {
         if (setState(ANIMATING)) {
             mClearAnimation = false
             val startTime = System.currentTimeMillis()
-            @Zoom val startZoom = zoom
-            @Zoom val endZoom = newZoom
+            @Zoom val startZoom = this.zoom
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1267]: I'm sorry - there are a number of small refactorings and variable renamings in this PR.
Android Studio was getting on my nerves with duplicate variable names (parameter and local variable with the same name) and I had a hard time understanding what the `time` variable actually is used for so I changed those things.
If there is anything that speaks against this I can try to reverse those changes - but if you want this please let me know exactly what I need to reverse.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 6]
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
                 }
 
                 // Having both overPinch and overScroll is hard to manage, there are lots of bugs if we do.
                 val factor = detector.scaleFactor
                 val newZoom = zoom * factor
-                applyPinch(newZoom, mAbsTargetX, mAbsTargetY, true)
+
+                // apply the new zoom and pan values accordingly
+                applyZoomAndAbsolutePan(newZoom,
+                        mCurrentAbsTargetX, mCurrentAbsTargetY,
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I just realized this pinch is missing the zoom target

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 7]
```diff
@@ -753,54 +764,154 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+                scaledFocusX = -detector.focusX
+                scaledFocusY = -detector.focusY
+                LOG.i("onScale:", "Setting focus.", "detectorFocusX:", scaledFocusX, "detectorFocusX:", scaledFocusY)
+
+                // Account for current pan.
+                scaledFocusX += scaledPanX
+                scaledFocusY += scaledPanY
+
+                mCurrentAbsTargetX = panX
+                mCurrentAbsTargetY = panY
```
Comment (markusressel) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: ~This should be nInitialAbsTargetX if I'm not mistaking. this might cause the little jump at the beginning when pinching fast.~

Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Forget this, I was mistaking.
I though about this pan shifting while pinching again and it's now fixed.
Its much more smooth, no initial jump and works in all situations I could come up with.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 6 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: fixed

File: library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt [hunk 8]
```diff
@@ -0,0 +1,49 @@
+package com.otaliastudios.zoom
+
+/**
+ * This class represents an absolute point on the ZoomEngine canvas (or beyond it's bounds)
+ */
+data class AbsolutePoint(
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt:6]: These make everything cleaner, thanks.
Are they exposed? If not let's make them internal.

File: library/src/test/java/com/otaliastudios/zoom/PointTests.kt [hunk 9]
```diff
@@ -0,0 +1,43 @@
+package com.otaliastudios.zoom
+
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.TestInstance
+
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Comment (natario1) [inline library/src/test/java/com/otaliastudios/zoom/PointTests.kt]: Thanks for tests!

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 10]
```diff
@@ -1235,34 +1356,81 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
         dispatchOnMatrix()
     }
 
+    /**
+     * Converts an [AbsolutePoint] to a [ScaledPoint]
+     */
+    private fun AbsolutePoint.toScaled(): ScaledPoint {
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1493]: Nice. I think we could finally remove `resolvePan` and `unresolvePan` and create extensions of the `Float` class with the same names?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 5 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1267]: I think it's fine

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 2 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:894]: I should take a better look to understand what's happening. Anyway I just tested the app and it's not super smooth so there must be a bug somewhere. 

For instance if I pinch out (move fingers closer) then release, it suddenly moves to top-left then goes back. It seems to do this during pinch+pan also. Do you have any idea?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 10 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1493]: I mean to call them toScaled() and toAbsolute() of course.

File: library/src/test/java/com/otaliastudios/zoom/PointTests.kt (hunk 9 above)
Comment (markusressel) [reply library/src/test/java/com/otaliastudios/zoom/PointTests.kt]: Well they are pretty basic so far, I just wanted to get the basic structure working. 

Initially I wanted to test methods of the engine but I failed to create a test that has a valid context that is needed by the engine :-(
I think this can be another PR though.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 2 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:894]: You are right when zooming out more than the size of the content (so there are gray areas all around the image) the view does jump when pinching out.
I will have a look at that.

I don't know of any other jumping though if this is not the scenario you are describing.

File: library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt (hunk 8 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt:6]: As `panX` and `panY` are exposed I also exposed `pan` which is an `AbsolutePoint`. I can. remove that if it should only be an internal representation but I think the easier handling of points can also be beneficial to devs using this library - don't you think?

Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/AbsolutePoint.kt:6]: Yes, cool!

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 10 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1493]: Sure.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 2 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:894]: I think I fixed it using the center of `mContentRect` as the pivot point when zooming in from underpinch. @natario1 Could you have another look?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 3 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I added a boolean parameter `triggerListeners` to the relevant methods to prevent triggering them when changes are only simulated.

File: app/build.gradle [hunk 11]
```diff
@@ -10,6 +10,8 @@ android {
         targetSdkVersion rootProject.ext.targetSdkVersion
         versionCode 1
         versionName "1.0"
+
+        setProperty("archivesBaseName", "ZoomLayout_v${versionName}_(${versionCode})")
```
Comment (natario1) [inline app/build.gradle:14]: What will this do?

File: library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt [hunk 12]
```diff
@@ -88,6 +91,7 @@ interface ZoomApi {
      * @see panY
      * @see AbsolutePan
      */
+    @Target(AnnotationTarget.FIELD, AnnotationTarget.FUNCTION, AnnotationTarget.LOCAL_VARIABLE, AnnotationTarget.PROPERTY, AnnotationTarget.VALUE_PARAMETER)
     @Retention(AnnotationRetention.SOURCE)
     annotation class ScaledPan
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:103]: I regret using these name, something like AbsoluteCoord would have been better. But we will change in another version

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 13]
```diff
@@ -150,9 +150,14 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
     @RealZoom
     override val realZoom: Float
-        @RealZoom
         get() = zoom * mBaseZoom
 
+    /**
+     * The current pan as an [AbsolutePoint]
+     */
+    private val pan: AbsolutePoint
+        get() = AbsolutePoint(panX, panY)
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Are these pan and scaledPan values accessed frequently?
If they are it might be worth to avoid all these new instances by caching the value (have a private _pan and use _pan.set() in this getter), since touch events are a lot.

By the way I thought from another comment that these two were public?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 14]
```diff
@@ -625,10 +637,25 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
     }
 
+    /**
+     * Uses [checkPanBounds] to calculate the pan correction that
+     * needs to be applied to get into a valid state.
+     *
+     * @return the scaled pan fix values for x- and y-axis
+     */
+    private fun getPanCorrection(): ScaledPoint {
+        return ScaledPoint(
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: See if it's better to avoid this new instance as well depending on how frequently this is accessed

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 2 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:894]: Seems to work great now! Nice work!

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 15]
```diff
@@ -773,58 +806,171 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+     * @return the currently set lower boundary for minimum zoom value
+     */
+    @Zoom
+    private fun getMinZoom(): Float = mMinZoom.toZoom(mMinZoomMode)
 
     private inner class FlingScrollListener : GestureDetector.SimpleOnGestureListener() {
 
         override fun onDown(e: MotionEvent): Boolean {
+            // cancel any animation that might still be running
+            setState(NONE)
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I see some issues with this: zoom in, then fling fast toward one end, then put your finger when the content is overscrolled. It does not always come back.

Might be easy to fix, but we can leave it for another PR I think.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 16]
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+        if (zoom <= getMinZoom()) {
+            zoomTo(getMinZoom(), animate = true)
         }
     }
 
     //endregion
 
     //region Apply values
 
+    private val mCancelAnimationListener = object : Animator.AnimatorListener {
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1232]: Do you think we should setState(NONE) onAnimationCancel also?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 17]
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+            @Zoom val endZoom = checkZoomBounds(zoom, allowOverPinch)
+            val zoomAnimator = ValueAnimator.ofFloat(startZoom, endZoom)
+            zoomAnimator.prepare()
+            zoomAnimator.addUpdateListener {
+                LOG.v("animateZoom:", "animationStep:", it.animatedFraction)
+                if (mClearAnimation) {
+                    it.cancel()
                 }
-            })
+                mContainer.postOnAnimation {
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: You don't need this post here, I think the animator already runs in an animation step.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 18]
```diff
@@ -1056,86 +1202,118 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+                    PropertyValuesHolder.ofFloat(
+                            "zoom",
+                            startZoom, endZoom)
+            )
+            animator.prepare()
+            animator.addUpdateListener {
+                if (mClearAnimation) {
+                    it.cancel()
+                }
+                mContainer.postOnAnimation {
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Same here, post is not needed. Nice approach

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 19]
```diff
@@ -1147,60 +1325,63 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
      * @param deltaY          a scaled delta
      * @param allowOverScroll whether to overscroll
      */
-    private fun animateScaledPan(@ScaledPan deltaX: Float, @ScaledPan deltaY: Float,
+    private fun animateScaledPan(@AbsolutePan deltaX: Float, @ScaledPan deltaY: Float,
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I think this was an error

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 20]
```diff
@@ -1147,60 +1325,63 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+                startValue + (endValue - startValue) * fraction - scaledPan
+            }, startPan, endPan)
+            panAnimator.prepare()
+            panAnimator.addUpdateListener {
+                LOG.v("animateScaledPan:", "animationStep:", it.animatedFraction)
+                if (mClearAnimation) {
+                    it.cancel()
                 }
-            })
+                mContainer.postOnAnimation {
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Same as before, post not needed

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 4 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: If it's unused we can remove it.

Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Ok

File: app/build.gradle (hunk 11 above)
Comment (markusressel) [reply app/build.gradle:14]: When building an apk (release or debug) the resulting apk file will be called this way.
I have too much `app-debug.apk` files in my life already :stuck_out_tongue:

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 13 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I made them public now, don't know why they weren't.
Yes using the same Object is probably better for performance - I changed that.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 19 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: jup good catch!

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 20 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: ok

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 18 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: ok

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 17 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: ok

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 14 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: done

File: library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt (hunk 12 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:103]: Those names still confuse me :grin: 
They are also kind of useless since the IDE seems to do nothing to check these.
Even when passing wrong values to a function with annotated parameters there is no warning.
That's another reason why I opted to create data objects for them (although you can still pass in wrong values on their constructors...)

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 15 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: ok I will just revert this little change in this PR then.

File: library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt (hunk 12 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:103]: Ah, without them it would be worse I think... There should be an android annotation to tell the IDE that these two annotations exclude each other, but I don't know how.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 16 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1232]: Well the animation is only cancelled when `mClearAnimation` is set to `true` which is done in `setState(NONE)` ... so currently it's the other way around - the state change causes the animationListener to trigger onCancel. It wouldn't hurt though I guess since it will already be in that state.

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 13 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: I didn't know you could do what you did :-) just one last change, can they be vals of ZoomApi overriden in the engine?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 16 above)
Comment (natario1) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1232]: Well it's not important then, as you wish

File: library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt (hunk 12 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomApi.kt:103]: Yes they do help indeed! But not to the extend I'd like them to :stuck_out_tongue:

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt (hunk 13 above)
Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt]: Well the `pan` val can, the `scaledPan` depends on `mTransformedRect` so I don't think so?

File: library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt [hunk 21]
```diff
@@ -1151,56 +1329,57 @@ internal constructor(context: Context) : ViewTreeObserver.OnGlobalLayoutListener
...
+            val endPan = startPan + ScaledPoint(deltaX, deltaY)
+
+            val panAnimator = ValueAnimator.ofObject(TypeEvaluator { fraction, startValue: ScaledPoint, endValue: ScaledPoint ->
+                startValue + (endValue - startValue) * fraction - scaledPan
+            }, startPan, endPan)
+            panAnimator.prepare()
+            panAnimator.addUpdateListener {
+                LOG.v("animateScaledPan:", "animationStep:", it.animatedFraction)
+                if (mClearAnimation) {
+                    it.cancel()
```
Comment (natario1) [inline library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1359]: Sorry I forgot this, we need a return after these it.cancel() to be as it was.

Comment (markusressel) [reply library/src/main/java/com/otaliastudios/zoom/ZoomEngine.kt:1359]: right, I changed it in all 3 places.

**Review by natario1 — APPROVED** [review]
This is good IMO! It was a great addition, thanks.

**Comment by markusressel** [comment]
Since you already approved this PR and I only fixed a small dev-merge mistake I will merge this PR now.
If any merge conflicts come up in #74 because of this and you are unsure what to do just let me know.
Cheers! :tada:
